# Proyecto integrador para el Módulo de analista de Datos - ISPC

#### Módulo: Analista de datos (Estádistica y Exploración de datos, Ciencia de Datos y Procesamiento de Datos)
#### Alumno: PALOMEQUE, Jonathan Manuel
#### Carrera: Tecnicatura Superior en Ciencia de datos e IA
#### Institución: Instituto Superior Politécnico Córdoba (ISPC)
#### Año: 2024

## Desarrollo del proyecto: 

### I) Importar Pandas, Numpy y Matplotlib:

In [46]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

### II) Cargar DataSet y crear el Dataframe:

In [47]:
dataSet = pd.read_csv('Ventas-Original.csv')

dataframe = pd.DataFrame(dataSet)

### III) Descripcion del DataSet:

**1- Informacion General sobre el dataframe:**

In [48]:
print(dataframe.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12014 entries, 0 to 12013
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   pais              12004 non-null  object 
 1   metodo_orden      12008 non-null  object 
 2   tipo_comercio     12005 non-null  object 
 3   linea             12004 non-null  object 
 4   tipo_producto     12008 non-null  object 
 5   producto          12005 non-null  object 
 6   año               12009 non-null  float64
 7   cuatrimestre      12014 non-null  object 
 8   ventas            12009 non-null  object 
 9   cantidad          12009 non-null  float64
 10  utilidad_bruta_p  12010 non-null  object 
 11  utilidad_bruta    12007 non-null  object 
 12  costos            12009 non-null  object 
dtypes: float64(2), object(11)
memory usage: 1.2+ MB
None


*El resultado brinda informacion de la cantidad de columnas, sus nombres, tipoa de datos y que hay 12.014 filas y 13 columnas. No hay valores Nulos en las columnas*

**2- Imprimir los primeros  10 registros del dataframe:**

In [49]:
print(dataframe.head(10))

        pais   metodo_orden      tipo_comercio     linea tipo_producto  \
0  Argentina  Representante          Mayorista  Mascotas      Juguetes   
1  Argentina           Mail      De proximidad  Mascotas      Juguetes   
2  Argentina       Teléfono  Grandes Almacenes  Infantil       Calzado   
3  Argentina        Website      Especializado    Hombre       Calzado   
4  Argentina        Website      Especializado   Adultos    Vestimenta   
5  Argentina        Website          Mayorista     Mujer        Lentes   
6  Argentina       Teléfono      Especializado     Mujer    Vestimenta   
7  Argentina  Representante       Hipermercado    Hombre        Lentes   
8  Argentina  Representante  Grandes Almacenes  Mascotas      Juguetes   
9  Argentina  Representante          Mayorista    Hombre       Calzado   

                 producto     año cuatrimestre       ventas  cantidad  \
0        Anitibyx armatus  2019.0      C1 2019  $155.040,02   15562.0   
1      Sarcorhamphus papa  2019.0      

**3- Cantidad de registros y columnas:**

In [50]:
cant_filas, cant_columnas = dataframe.shape

print("""
La Cantidad de registros son: {}
La Cantidad de columnas son: {}""".format(cant_filas, cant_columnas))


La Cantidad de registros son: 12014
La Cantidad de columnas son: 13


**4- Cantidad de elementos que NO son nulos por columnas:**

In [51]:
dataframe.count()

pais                12004
metodo_orden        12008
tipo_comercio       12005
linea               12004
tipo_producto       12008
producto            12005
año                 12009
cuatrimestre        12014
ventas              12009
cantidad            12009
utilidad_bruta_p    12010
utilidad_bruta      12007
costos              12009
dtype: int64

**5-Tipos de Datos:**

In [52]:
tipos_datos = dataframe.dtypes
print(tipos_datos)

pais                 object
metodo_orden         object
tipo_comercio        object
linea                object
tipo_producto        object
producto             object
año                 float64
cuatrimestre         object
ventas               object
cantidad            float64
utilidad_bruta_p     object
utilidad_bruta       object
costos               object
dtype: object


*Verifico que hay columnas cuyos datos deberian ser del tipo entero o flotante, los cuales actualmente son del tipo objeto, por lo que deben ser modificados para poder realizar el analisis de los mismos*

### IV) Limpieza y Procesamiento de datos:

**1- Datos en Formato incorrecto:**

Es necesario cambiar el tipo de dato de las columnas: *ventas, utilidad_bruta,  costos y utilidad_bruta_p*, del tipo objeto al tipo flotante, ya que verifico que los registros de las columnas ventas, utilidad_bruta y costos, originalmente tiene la siguiente sintaxis: "$155.040,02" y necesito que queden de la siguiente forma: "155040.02" , para poder obtener las medidas de tendencias centrales (media, mediana y desviacion estandar) asi como para realizar acciones de conteo. 

Para hacerlo seguiré los siguientes pasos:
- eliminar los signos "$"
- eliminar as ","
- reemplazar los "." por "," (Ya que sino, no puedo cambiar el tipo de dato)

**A- Reemplazar los signos "$" de las columnas: ventas, 'utilidad_bruta' y 'costos'**

In [53]:
dataframe['ventas'] = dataframe['ventas'].str.replace('$', '')
dataframe['utilidad_bruta'] = dataframe['utilidad_bruta'].str.replace('$', '')
dataframe['costos'] = dataframe['costos'].str.replace('$', '')

**B- Reemplazar los signos "." de las columnas: ventas, 'utilidad_bruta', 'costos' y 'cantidad'**

In [54]:
dataframe['ventas'] = dataframe['ventas'].str.replace('.', '')
dataframe['utilidad_bruta'] = dataframe['utilidad_bruta'].str.replace('.', '')
dataframe['costos'] = dataframe['costos'].str.replace('.', '')

**C - Reemplazar los signos "," por "." de las columnas:ventas, , 'utilidad_bruta', 'costos' y 'cantidad'**

In [55]:
dataframe['ventas'] = dataframe['ventas'].str.replace(',', '.')
dataframe['utilidad_bruta'] = dataframe['utilidad_bruta'].str.replace(',', '.')
dataframe['costos'] = dataframe['costos'].str.replace(',', '.')

**D- Reemplazar los signos "%" por "" y "." por "," de la columna "utilidad_bruta_p"**

In [56]:
dataframe['utilidad_bruta_p'] = dataframe['utilidad_bruta_p'].str.replace('%','')
dataframe['utilidad_bruta_p'] = dataframe['utilidad_bruta_p'].str.replace(',','.')

**E- Ahora si, puedo convertir cambiar el tipo de dato a float:**

In [57]:
dataframe[['ventas', 'utilidad_bruta', 'costos', 'utilidad_bruta_p']] = dataframe[['ventas', 'utilidad_bruta', 'costos', 'utilidad_bruta_p']].astype(float)

**F-Tipos de datos finales:**

In [58]:
tipos_datos = dataframe.dtypes
print(tipos_datos)

pais                 object
metodo_orden         object
tipo_comercio        object
linea                object
tipo_producto        object
producto             object
año                 float64
cuatrimestre         object
ventas              float64
cantidad            float64
utilidad_bruta_p    float64
utilidad_bruta      float64
costos              float64
dtype: object


**2- Limpieza de celdas vacias:**

*Utilizo el metodo  dropna() para eliminar las filas que contengan celdas vacias, como deseo conservar el DataFrame original, utilizo el argumento inplace=True*

In [59]:
dataframe.dropna(inplace = True)

*Verifico la cantidad de registros  resultantes luego de la limpieza:*

In [60]:
ncant_filas, ncant_columnas = dataframe.shape
print(f"Cantidad de registros: {ncant_filas}. Cantidad de columnas: {ncant_columnas}.")

Cantidad de registros: 11952. Cantidad de columnas: 13.


*De un total original de 12014 filas, quedaron 11952, lo que nos indica que habia 62 filas con celdas vacias*

**3- Limpieza de filas duplicadas:**

**A- Identificar las filas duplicadas:**

*Uso el metodo duplicated() para identificar duplicados, como este metodo nos devuelve True si las filas estan duplicadas, no es practico de visualizar cuando son muchos registros, por lo que primero guardo el resultado en un nuevo dataset (filasRegistro), para luego recorrer cada celda  y guardar en un array el indice de aquellas que son  True (duplicados), asi poder llevar un control de cuales son las duplicadas.*

In [61]:
filasRegistro = dataframe.duplicated()
filasDuplicadas = []

for x in filasRegistro.index:
  if filasRegistro.loc[x] == True:
    filasDuplicadas.append(x)

if len(filasDuplicadas) == 0:
    print('No hay filas duplicadas')
else:
    print('Las filas duplicadas son: {}'.format(filasDuplicadas))

Las filas duplicadas son: [11, 1973, 2001, 2009, 3499, 7091, 7871, 7908, 9269, 9301, 9677, 9679, 11750, 11778]


**B - Eliminar filas duplicadas:**

In [64]:
dataframe.drop_duplicates(inplace = True)

*Verifico la cantidad de registros resultantes luego de la limpieza:*

In [65]:
cant_filas, cant_columnas = dataframe.shape
print(f"Cantidad de registros resultantes: {cant_filas}. Cantidad de columnas: {cant_columnas}")

Cantidad de registros resultantes: 11938. Cantidad de columnas: 13


*De un total original de 12.014 filas, pasamos a 11.952 y ahora nos quedan 11.938 lo que nos indica que habia 14 filas con celdas duplicadas*

### V) Estadísticas:

**1- Informacion estadistica del dataset:**

In [67]:
infoEstadistica = dataframe.describe()
infoEstadistica.round(2)

año     ventas  cantidad  utilidad_bruta_p  utilidad_bruta  \
count   11938.00   11938.00  11938.00          11938.00        11938.00   
mean     2184.32  137962.21   9001.98             49.90        68879.06   
std      5227.81   27455.12   4911.78             14.78        24821.31   
min      2019.00       0.00  -9049.00             24.58        22407.08   
25%      2019.00  114467.48   4750.25             37.16        49138.76   
50%      2020.00  137784.54   8998.50             49.72        66374.43   
75%      2021.00  161712.22  13243.25             62.89        85773.92   
max    201900.00  185644.24  17521.00             75.40       138447.51   

          costos  
count   11938.00  
mean    69148.05  
std     24830.83  
min     22339.56  
25%     49477.16  
50%     66511.94  
75%     86449.10  
max    139072.81

**2- Media:**

In [70]:
# media de Ventas:
media_ventas = dataframe['ventas'].mean()

# media de Cantidad de productos:
media_cant_productos = dataframe['cantidad'].mean()

# media de Utilidad Bruta:
media_utilidad_bruta = dataframe['utilidad_bruta'].mean()

# media de costos:
media_costos =  dataframe['costos'].mean()


print("""
La media de ventas es: ${} 
La media de la cantidad de productos es: {} 
La media de utilidades brutas es: ${} 
La media de costos es: ${}
""".format(media_ventas.round(2), media_cant_productos.round(2), media_utilidad_bruta.round(2), media_costos.round(2)))


La media de ventas es: $137962.21 
La media de la cantidad de productos es: 9001.98 
La media de utilidades brutas es: $68879.06 
La media de costos es: $69148.05



**3- Mediana:**

In [71]:
# mediana de Ventas:
mediana_ventas = dataframe['ventas'].median()

# mediana de Cantidad de productos:
mediana_cant_productos = dataframe['cantidad'].median()

# mediana de Utilidad Bruta:
mediana_utilidad_bruta = dataframe['utilidad_bruta'].median()

# mediana de costos:
mediana_costos =  dataframe['costos'].median()


print("""
La mediana de ventas es: ${} 
La mediana de la cantidad de productos es: {} 
La mediana de utilidades brutas es: ${} 
La mediana de costos es: ${}
""".format(mediana_ventas, mediana_cant_productos, mediana_utilidad_bruta, mediana_costos))


La mediana de ventas es: $137784.53999999998 
La mediana de la cantidad de productos es: 8998.5 
La mediana de utilidades brutas es: $66374.43 
La mediana de costos es: $66511.94



**4- Desviacion Estandar:**

In [73]:
# Desviacion estándar en ventas:
desviacion_ventas = dataframe['ventas'].std()

#Desviación estándar de Cantidad de Productos:
desviacion_cant_productos = dataframe['cantidad'].std()

#Desviación estándar de utilidad bruta:
desviacion_utilidad_bruta = dataframe['utilidad_bruta'].std()

#Desviacion estándar de costos:
desviacion_costos = dataframe['costos'].std()

print("""
La desviación estándar de ventas es: ${}
La Desviación estándar de cantidad de productos es: {}
La desviación estándar de utilidad bruta es: {}
La desviacion estandar de costos es: {}""".format(desviacion_ventas, desviacion_cant_productos, desviacion_utilidad_bruta, desviacion_costos))


La desviación estándar de ventas es: $27455.12323591594
La Desviación estándar de cantidad de productos es: 4911.7777816223315
La desviación estándar de utilidad bruta es: 24821.312374768153
La desviacion estandar de costos es: 24830.833115365098
